### PR TITLE
Port Intet functionality

### PR DESCRIPTION
**Pull request by nitbhat**
Original location of the `intet` function: https://github.com/quinoacomputing/quinoa/blob/develop/src/Mesh/DerivedData.cpp#L1695

---
This change is [Reviewable](https://reviewable.io/reviews/charmworks/exam2m/7)

### CODE REVIEW

File: src/Transfer/Worker.cpp [hunk 1]
```diff
@@ -212,21 +212,124 @@ Worker::processCollisions(
...
+  for(int i = 0; i < nPoints; i++) {
+    bool val = intet(points[i].first,
+                     points[i].second,
+                     N);
+
+    if(val)
+      numInTet++;
+
+    // TODO: I'm guessing we need the point(points[i]), the tet index(points[i].second), and the shapefunction(N) to be stored from this function for future use?
+  }
```
Comment (jbakosi) [inline src/Transfer/Worker.cpp:246]: Yes, we will use the finite element shape functions, evaluated at the point location, to interpolate the results between meshes. (The points/vertices of the destination mesh will fall into tets of the source mesh whose vertices will hold the current solution to be interpolated.)

File: src/Transfer/Worker.cpp [hunk 2]
```diff
@@ -212,21 +212,124 @@ Worker::processCollisions(
   }
 
   for (int i = 0; i < numchares; i++) {
-    proxy[i].checkPoints(separated[i].size(), separated[i].data());
+    proxy[i].determineActualCollisions(separated[i].size(), separated[i].data());
```
Comment (jbakosi) [inline src/Transfer/Worker.cpp:216]: If I understand correctly, this is currently a one-to-all spawned from a loop. This is good to get things started and correct and we will switch this to a broadcast (one-to-all), and ultimately, a broadcast to a subset of potential workers, e.g., those that had destination mesh vertices found in the source mesh in the previous time step.

**Review by jbakosi — APPROVED** [review]

**Comment by nitbhat** [comment]
Currently, it looks like the real collisions (as detected by the intet function) are a very small fraction of the possible collisions. 


```
1: Worker 0 received data for 22745 collisions
1: [0]: Received data for 20030 potential collisions with my tets
1: [0]: 27 collisions are actually in my tets out of 20030 potential collisions
1: Worker 1 received data for 22950 collisions
1: [1]: Received data for 2715 potential collisions with my tets
1: [1]: 15 collisions are actually in my tets out of 2715 potential collisions
1: [1]: Received data for 20928 potential collisions with my tets
1: [0]: Received data for 2022 potential collisions with my tets
1: [0]: 4 collisions are actually in my tets out of 2022 potential collisions
1: [1]: 34 collisions are actually in my tets out of 20928 potential collisions
1: Quiescence detected
1: [Partition 0][Node 0] End of program
```

@jbakosi: Does this result look correct? Are we missing something?

**Comment by jbakosi** [comment]
What two meshes are you running and how are they oriented wrt each other?

**Comment by nitbhat** [comment]
This is the output of the CI testing, which runs the command `mpirun -np 2  ./Main/exam2m ../tests/sphere_tetra.0.2.exo ../tests/sphere_tetra.0.2.exo`, so it's the `sphere_tetra.0.2.exo` mesh tested for collisions with itself.

**Comment by jbakosi** [comment]
Oh. I'm not sure how `intet()` will behave in such a case. I agree that a self-collision is a good one to test, but I'm not sure how the interpolation/intersection algorithm in `intet()` will (supposed to) find all vertices of a mesh _within_ some tets of the same mesh. This is a pretty pathological case, considering that the interpolation performs a solution of a small linear system, i.e., a bunch of floating point arithmetic. In other words, I would not trust it too much when the point is very close (i.e., in the floating point distance sense) to a vertex of the source mesh tet.

At first, I would test this with a less pathological case, e.g., the small sphere inside the large sphere, and make sure that that finds all intersections it is supposed to.

**Comment by nitbhat** [comment]
Okay, thanks Jozsef, that sounds good. I'll try determining that test result and post it here.

**Comment by nitbhat** [comment]
It looks like the collision detection library doesn't find any collisions for that case. 

```
nbhat4@courage:/scratch/nitin/ExaM2M/build$ mpirun -n 2 ./Main/exam2m ../tests/sphere_tetra.2.exo ../tests/sphere_tetra.0.2.exo
Charm++> Running on MPI version: 3.0
Charm++> level of thread support used: MPI_THREAD_SINGLE (desired: MPI_THREAD_SINGLE)
Charm++> Running in non-SMP mode: 2 processes (PEs)
Converse/Charm++ Commit ID:
Isomalloc> Synchronized global address space.
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
Charm++> cpu topology info is gathered in 0.000 seconds.
Args: ../tests/sphere_tetra.2.exo ../tests/sphere_tetra.0.2.exo
Initial load distribution
Virtualization [0.0...1.0]: 0
Number of work units: 2
ZOLTAN Load balancing method = 3 (RCB)

  Build configuration:
  ZOLTAN_ID_TYPE: unsigned int (4 bytes)
  ZOLTAN_GNO_TYPE: ssize_t, (8 bytes)
  MPI_Datatype for ZOLTAN_ID_TYPE: MPI_UNSIGNED
  MPI_Datatype for ZOLTAN_GNO_TYPE: MPI_LONG
ZOLTAN Parameter IMBALANCE_TOL[0] = 1.100000
ZOLTAN Parameter AUTO_MIGRATE = FALSE
ZOLTAN Parameter MIGRATE_ONLY_PROC_CHANGES = 1
ZOLTAN Parameter OBJ_WEIGHT_DIM = 0
ZOLTAN Parameter EDGE_WEIGHT_DIM = 0
ZOLTAN Parameter DEBUG_LEVEL = 1
ZOLTAN Parameter DEBUG_PROCESSOR = 0
ZOLTAN Parameter DETERMINISTIC = TRUE
ZOLTAN Parameter TIMER = 1 (wall)
ZOLTAN Parameter NUM_GID_ENTRIES = 2
ZOLTAN Parameter NUM_LID_ENTRIES = 2
ZOLTAN Parameter RETURN_LISTS = PARTITION ASSIGNMENTS
ZOLTAN Parameter NUM_GLOBAL_PARTS = 2
ZOLTAN Parameter NUM_LOCAL_PARTS = -1
ZOLTAN Parameter REMAP = 1
ZOLTAN Parameter SEED = 123456789 (123456789)
ZOLTAN Parameter LB_APPROACH = repartition
ZOLTAN Parameter RCB_OVERALLOC = 1.200000
ZOLTAN Parameter RCB_REUSE = 0
ZOLTAN Parameter CHECK_GEOM = 1
ZOLTAN Parameter RCB_OUTPUT_LEVEL = 0
ZOLTAN Parameter KEEP_CUTS = 0
ZOLTAN Parameter RCB_LOCK_DIRECTIONS = 0
ZOLTAN Parameter RCB_SET_DIRECTIONS = 0
ZOLTAN Parameter RCB_RECTILINEAR_BLOCKS = 0
ZOLTAN Parameter OBJ_WEIGHTS_COMPARABLE = 0
ZOLTAN Parameter RCB_MULTICRITERIA_NORM = 1
ZOLTAN Parameter RCB_MAX_ASPECT_RATIO = 10.000000
ZOLTAN Parameter AVERAGE_CUTS = 0
ZOLTAN Parameter RANDOM_PIVOTS = 0
ZOLTAN Parameter RCB_RECOMPUTE_BOX = 0
ZOLTAN Parameter REDUCE_DIMENSIONS = 0
ZOLTAN Parameter DEGENERATE_RATIO = 0.000000
ZOLTAN Parameter FINAL_OUTPUT = 0
ZOLTAN Parameter KEEP_CUTS = 0
ZOLTAN Parameter REDUCE_DIMENSIONS = 0
ZOLTAN Parameter DEGENERATE_RATIO = 10.000000
Initial load distribution
Virtualization [0.0...1.0]: 0
Number of work units: 2
ZOLTAN Load balancing method = 3 (RCB)

  Build configuration:
  ZOLTAN_ID_TYPE: unsigned int (4 bytes)
  ZOLTAN_GNO_TYPE: ssize_t, (8 bytes)
  MPI_Datatype for ZOLTAN_ID_TYPE: MPI_UNSIGNED
  MPI_Datatype for ZOLTAN_GNO_TYPE: MPI_LONG
ZOLTAN Parameter IMBALANCE_TOL[0] = 1.100000
ZOLTAN Parameter AUTO_MIGRATE = FALSE
ZOLTAN Parameter MIGRATE_ONLY_PROC_CHANGES = 1
ZOLTAN Parameter OBJ_WEIGHT_DIM = 0
ZOLTAN Parameter EDGE_WEIGHT_DIM = 0
ZOLTAN Parameter DEBUG_LEVEL = 1
ZOLTAN Parameter DEBUG_PROCESSOR = 0
ZOLTAN Parameter DETERMINISTIC = TRUE
ZOLTAN Parameter TIMER = 1 (wall)
ZOLTAN Parameter NUM_GID_ENTRIES = 2
ZOLTAN Parameter NUM_LID_ENTRIES = 2
ZOLTAN Parameter RETURN_LISTS = PARTITION ASSIGNMENTS
ZOLTAN Parameter NUM_GLOBAL_PARTS = 2
ZOLTAN Parameter NUM_LOCAL_PARTS = -1
ZOLTAN Parameter REMAP = 1
ZOLTAN Parameter SEED = 460185016 (460185016)
ZOLTAN Parameter LB_APPROACH = repartition
ZOLTAN Parameter RCB_OVERALLOC = 1.200000
ZOLTAN Parameter RCB_REUSE = 0
ZOLTAN Parameter CHECK_GEOM = 1
ZOLTAN Parameter RCB_OUTPUT_LEVEL = 0
ZOLTAN Parameter KEEP_CUTS = 0
ZOLTAN Parameter RCB_LOCK_DIRECTIONS = 0
ZOLTAN Parameter RCB_SET_DIRECTIONS = 0
ZOLTAN Parameter RCB_RECTILINEAR_BLOCKS = 0
ZOLTAN Parameter OBJ_WEIGHTS_COMPARABLE = 0
ZOLTAN Parameter RCB_MULTICRITERIA_NORM = 1
ZOLTAN Parameter RCB_MAX_ASPECT_RATIO = 10.000000
ZOLTAN Parameter AVERAGE_CUTS = 0
ZOLTAN Parameter RANDOM_PIVOTS = 0
ZOLTAN Parameter RCB_RECOMPUTE_BOX = 0
ZOLTAN Parameter REDUCE_DIMENSIONS = 0
ZOLTAN Parameter DEGENERATE_RATIO = 0.000000
ZOLTAN Parameter FINAL_OUTPUT = 0
ZOLTAN Parameter KEEP_CUTS = 0
ZOLTAN Parameter REDUCE_DIMENSIONS = 0
ZOLTAN Parameter DEGENERATE_RATIO = 10.000000
Number of tetrahedra for mesh 0: 25920
Number of points for mesh 0: 4578
Number of tetrahedra for mesh 1: 25920
Number of points for mesh 1: 4578
Collisions found: 0
Source mesh chunk 0 has 0
Source mesh chunk 1 has 0
Worker 0 received data for 0 collisions
[0]: Received data for 0 potential collisions with my tets
[0]: 0 collisions are actually in my tets out of 0 potential collisions
Worker 1 received data for 0 collisions
[0]: Received data for 0 potential collisions with my tets
[0]: 0 collisions are actually in my tets out of 0 potential collisions
Quiescence detected
[1]: Received data for 0 potential collisions with my tets
[1]: 0 collisions are actually in my tets out of 0 potential collisions
[1]: Received data for 0 potential collisions with my tets
[1]: 0 collisions are actually in my tets out of 0 potential collisions
[Partition 0][Node 0] End of program
```

**Comment by nitbhat** [comment]
@jbakosi: No collisions are reported for the small sphere in the large sphere test case as indicated in the output above. Are they expected to have collisions?

**Comment by jbakosi** [comment]
I cannot remember but the large sphere has a  sphere hole cut out in the center. Depending on the hole diameter, the small sphere might fall into it. You can visualize them with paraview.

**Comment by jbakosi** [comment]
You can also scale the small sphere mesh by simply multiplying the coordinate array with a number larger than one if you need it larger because it is centered at the origin.

**Pull request merged**